modules/writing_observer/writing_observer/awe_nlp.py

```diff
@@ -202,52 +204,104 @@ async def process_texts_parallel(texts, options=None):
-    4. If we need to process anything process it
-    5. Update results/cache
+    Caching:
+    1. Create text hash.
+    2. Check if hash exist in cache.
+    3. Check if some options are a subset of features_available
+        Yes: add the intersection of features_available and options to results
+    4. Check if some options are a subset of features_running.
+        Yes: a. Wait for features_running to finish.
+             b. Update the cache
+             c. Add intersection of features_running and Options to results
+    5. Check if additional features are required.
+        Yes: a. Collect options not covered till now and add to features_running.
+             b. Once finished, update cache and return results.
     '''
+    recurring_sleep_time = 1  # Time to wait between recurring calls to cache to check if features have finished running
+    running_features_wait_time = 60  # Time to wait for features already running.
+    results = []
+    found_options = set()
+    cache = learning_observer.kvs.KVS()
+    if options is None:
+        options = []
+    set_options = set(options)
     processor = {
         RUN_MODES.MULTIPROCESSING: process_texts_parallel,
         RUN_MODES.SERIAL: process_texts_serial
     }
 
-    results = []
-    need_processing = []
-    needed_options = set()
-    cache = learning_observer.kvs.KVS()
-
     for writing in writing_data:
         text = writing.get('text', '')
         if len(text) == 0:
             continue
+
+        # Creating text hash
         text_hash = 'NLP_CACHE_' + learning_observer.util.secure_hash(text.encode('utf-8'))
         text_cache_data = await cache[text_hash]
+
         if text_cache_data is None:
             text_cache_data = {}
-        writing.update(text_cache_data)
-        missing_options = set(options if options is not None else []).difference(text_cache_data.keys())
-        needed_options.update(missing_options)
-        if len(missing_options) == 0:
-            results.append(writing)
+
+        if 'features_running' not in text_cache_data:
+            features_running = set()
         else:
-            need_processing.append(writing)
-
-    if len(need_processing) > 0:
-        just_the_text = [w.get("text", "") for w in need_processing]
-        annotated_texts = await processor[mode](just_the_text, list(needed_options))
-
-        for annotated_text, single_doc in zip(annotated_texts, need_processing):
-            if annotated_text != "Error":
-                single_doc.update(annotated_text)
-                text_hash = 'NLP_CACHE_' + learning_observer.util.secure_hash(single_doc['text'].encode('utf-8'))
-                new_cache = {k: v for k, v in single_doc.items() if k != 'student'}
-                await cache.set(text_hash, new_cache)
-        results.extend(need_processing)
+            features_running = set(json.loads(text_cache_data['features_running']))
+
+        if 'features_available' not in text_cache_data:
```

Comment:
instead of checking if each of these items are in the text_cache_data, would it better to set defaults in the text_cache_data?

It feels weird to constantly check a conditional, when we could just set a default instead.